### ISSUE TITLE
Files deleted after processing

### ISSUE THREAD

**Issue by Sappharad** [Sappharad/mp3gainOSX#22]
A user sent me a video recording of them trying to use MP3Gain Express to process a bunch of files. As soon as each file finishes processing, it disappears from their hard drive and cannot be found via search in the finder as if it had been deleted. The user claimed antivirus was not the cause. Nothing seemed unusual about their environment from the video. The files would process normally and the result would come back from MP3Gain, but disappear as soon as they were closed after writing the changes. There was never a situation where the program would ever work on the user's machine.

Only one person has ever reported running into this problem. Reported February 2019. I have no idea how to reproduce this locally or any thoughts on a cause. It will probably remain an outstanding problem until it happens to someone else and they figure out what's different about their machine that causes it.

Ultimately the problem may lie in the original command line MP3Gain application, but the user who reported it wasn't familiar with the terminal so I couldn't have them confirm the problem against the original tool.

**Comment by Sixten66** [comment]
I can confirm the same problem running MP3Gain Express 3.2.0 on macOS Mojave 10.14.6.
I don't think the problem is related to machine specifications, but it has something to do with certain files.
It processes normally some files while it deletes others. It deletes those files even when only analyzing them, without applying any gain changes.
I'm sending you two files to test it for yourself. In my case the file _09 Through The Barricades.m4a_ is being processed normally, while the file _02 Nothing's Gonna Stop Us Now.m4a_ is being deleted.
[Examples.zip](https://github.com/Sappharad/mp3gainOSX/files/3509709/Examples.zip)

**Comment by Sappharad** [comment]
@Sixten66 Wow, thanks so much for providing samples.

The one thing I missed in this entire ordeal is the file format - The person who initially reported it sent me a video recorded with their phone and the text was out of focus. Watching it again, the file extensions were definitely m4a and not mp3. Although MP3Gain is stable and I've never had issues with it, there is a big all caps warning on the official AACGain page about backing up your files:
http://aacgain.altosdesign.com/
But his reason involves the undo feature, not files completely going away after processing.

When the problem came in, I reviewed the MP3Gain code and didn't see any possible way that could happen. I never thought about AAC, because I encode all of my music to highest quality VBR MP3. Sadly, AACGain is significantly more complex than MP3Gain so this might not be a fun one to track down. I will try to look into it in the next few weeks.

**Comment by Sappharad** [comment]
One thing to point out about your files, _02 Nothing's Gonna Stop Us Now.m4a_ appears to have been purchased from iTunes and has account information embedded. Although the music is DRM free, they still put user info inside in the file to track piracy.

The other file doesn't have iTunes information embedded. I'm not sure if this is related, but something to keep in mind. I edited your post and removed the files, just in case you weren't aware of that. I have them now for future analysis. I'll have to test it on my own iTunes purchases later, I've never had to use it on those because usually they're already a good volume.

**Comment by Sixten66** [comment]
Both files are purchased from iTunes using the same account. Moreover, they are part of the same iTunes compilation (I later change the info to reflect the original recordings). So they both should have iTunes information embedded.
I'm quite sure the same problem occurs also with MP3 files, but I will have to search and test to find some files.
The same problem occurs with the original MP3Gain (running under WINE or natively on Windows). The only difference is that the original software deletes the files leaving a useless .tmp file. It does so even if I choose '_Do not use Temp files_' under the _Advanced Options_.
Before finding your thread here I have tried to find the reason for this problem myself checking out if some other process changes the file structure and produces the problem. These processes include changes made with _iVolume_, _Mixed in Key_ and changing the file's ID3 tag info, but none of these seems to have any affect.
Fortunately, I only make changes to backed up files, so I never have to worry about destroying or losing any files.

<img width="812" alt="Screenshot 2019-08-20 at 2 09 11 PM" src="https://user-images.githubusercontent.com/54177176/63342657-b9991f80-c354-11e9-9b20-9858b13d2ce2.png">
<img width="812" alt="Screenshot 2019-08-20 at 2 09 33 PM" src="https://user-images.githubusercontent.com/54177176/63342658-ba31b600-c354-11e9-9b32-7fcb4c400ef9.png">

**Comment by Sappharad** [comment]
I've tracked down the cause of the file deletion.

AACGain applies changes to a temporary copy of the file, and once that's done it makes a second "optimized" copy of the first temporary copy which removes some extra data that it adds while processing. The problem is that the AACgain code never expects optimize to fail and doesn't check the return value to see if it was successful. It assumes the second temp file got created so it removes the original file and tries to replace it with the optimized output, but that doesn't actually exist so it fails. Unfortunately, the first temp file appears to be corrupt in the first place, so we can't even salvage that. 

In other words, AACGain can't handle these files that are getting deleted. I'm not knowledgeable enough about the inner workings of AAC (mp4) files to fix that part, but I can fix the delete problem and let MP3Gain Express know the file failed to process.

The problem originates in the original AACgain code, so I've forked AACGain and committed the fix. I will try to get a new release of MP3Gain Express out before the end of the year so that at least it can properly fail on these files instead of delete them.
https://github.com/Sappharad/aacgain/commit/dec0364c681bbeed5854be5d85d930059fd0fbc9

**Comment by Sappharad** [comment]
The next release will flag incompatible files as not compatible and fix the delete problem.
<img width="627" alt="Screen Shot 2019-12-15 at 8 03 20 PM" src="https://user-images.githubusercontent.com/1350664/70874068-28d6f700-1f76-11ea-8778-ebe0914c1b23.png">

If aacgain is ever updated to make them compatible, then I'll upgrade to that. It's unlikely since it hasn't gotten an official update in years.
